src/nvim/api/vim.c

```diff
@@ -2389,3 +2389,151 @@ Dictionary nvim_complete_set(Integer index, Dict(complete_set) *opts, Arena *are
   }
   return rv;
 }
+
+enum { kStatusColumn = 0x1, kStatusLine = 0x2, kTabLine = 0x4, kWinBar = 0x8, };
+
+static void redraw_status(win_T *wp, buf_T *buf, int status, bool *flush)
+{
+  if (buf != NULL && wp->w_buffer != buf) {
+    return;
+  }
+
+  if (*wp->w_p_stc != NUL && (status & kStatusColumn)) {
+    wp->w_nrwidth_line_count = 0;
+    changed_window_setting(wp);
+  }
+  win_grid_alloc(wp);
+
+  // Just hid or showing winbar for the first time, or drawing the statuscolumn.
+  // Flush later because the window itself should be redrawn.
+  if (wp->w_lines_valid == 0) {
+    *flush = true;
+  }
+
+  if (*flush && (status & (kWinBar|kStatusLine))) {
+    // When flushing later, just mark for redraw.
+    wp->w_redr_status = true;
+  } else if (status & (kWinBar|kStatusLine)) {
+    // Not flushing later, only redraw the 'winbar' or 'statusline'.
+    win_check_ns_hl(wp);
+    if (status & kWinBar) {
+      win_redr_winbar(wp);
+    }
+    if (status & kStatusLine) {
+      win_redr_status(wp);
+    }
+    win_check_ns_hl(NULL);
+  }
+}
+
+/// Instruct Nvim to redraw various components.
+///
+/// @param opts  Optional parameters.
+///               - win: Target a specific |window-ID| for various components.
+///               - buf: Target a specific buffer number for various
+///                 components.
+///               - flush: Update the screen with pending updates, even if
+///                 Nvim is waiting for input and would normally not do so.
+///               - clear: Forcefully redraw `win`, all windows containing
+///                 `buf`, or all windows.
+///               - range: Buffer range to redraw, useful for decoration
+///                 providers. Expects a tuple `[first, last]` with the first
+///                 and last line number of the range. Targets `buf`, the
+///                 buffer in `win` or the current buffer.
+///               - cursor: Immediately update cursor position on the screen
+///                 in window `win` or the current window.
+///               - statuscolumn: Redraw the |'statuscolumn'| in `buf`, `win`
+///                 or all windows.
+///               - statusline: Redraw the |'statusline'| in `buf`, `win` or
+///                 all windows.
+///               - winbar: Redraw the |'winbar'| in `buf`, `win` or all
+///                 windows.
+///               - tabline: Redraw the |'tabline'|.
+void nvim_redraw(Dict(redraw) *opts, Error *err)
+  FUNC_API_SINCE(12)
+{
+  win_T *win = NULL;
+  buf_T *buf = NULL;
+  bool save_lz = p_lz;
+  int save_rd = RedrawingDisabled;
+  RedrawingDisabled = 0;
+  p_lz = false;
+
+  if (HAS_KEY(opts, redraw, win)) {
+    win = find_window_by_handle(opts->win, err);
+    if (ERROR_SET(err)) {
+      goto theend;
+    }
+  }
+
+  if (HAS_KEY(opts, redraw, buf)) {
+    VALIDATE(win == NULL, "%s", "Invalid 'buf': already specified 'win'.", {
```

Comment:
API error messages usually don't end with a period
```suggestion
    VALIDATE(win == NULL, "%s", "Invalid 'buf': already specified 'win'", {
```